# 012-P01. Weekly Study Summary

*Store a synthetic log in collections, apply Boolean rules in loops, and return a summary.*

Read the problem statement and the following theory sections before running the notebook: Boolean Rules and Branching, Collections and Iteration, Functions and Object State.

## Case Data

The problem supplies five synthetic records with `day`, `subject`, `hours`, and `completed` fields. The first output preserves all records before aggregation.

## Method Application · Source Record Inspection

The input table shows row order, repeated subjects, durations, and Boolean completion values.

In [1]:
study_records = [
    {"day": "Mon", "subject": "Python", "hours": 1.5, "completed": True},
    {"day": "Tue", "subject": "Statistics", "hours": 2.0, "completed": True},
    {"day": "Wed", "subject": "Python", "hours": 1.0, "completed": False},
    {"day": "Thu", "subject": "Data", "hours": 2.5, "completed": True},
    {"day": "Fri", "subject": "Python", "hours": 2.0, "completed": True},
]

print("Day | Subject | Hours | Completed")
for record in study_records:
    print(
        record["day"], "|", record["subject"], "|",
        record["hours"], "|", record["completed"]
    )

Day | Subject | Hours | Completed
Mon | Python | 1.5 | True
Tue | Statistics | 2.0 | True
Wed | Python | 1.0 | False
Thu | Data | 2.5 | True
Fri | Python | 2.0 | True


## Method Application · Subject-Hour Accumulation

The loop accumulates `hours` under each `subject` label without using optional compact syntax.

In [2]:
hours_by_subject = {}

for record in study_records:
    subject = record["subject"]
    if subject not in hours_by_subject:
        hours_by_subject[subject] = 0.0
    hours_by_subject[subject] = hours_by_subject[subject] + record["hours"]

print("Hours by subject:", hours_by_subject)

Hours by subject: {'Python': 4.5, 'Statistics': 2.0, 'Data': 2.5}


## Method Application · Reusable Weekly Summary

The function returns overall counts and duration in one labeled dictionary.

In [3]:
def summarize_week(records):
    total_hours = 0.0
    completed_sessions = 0
    for record in records:
        total_hours = total_hours + record["hours"]
        if record["completed"]:
            completed_sessions = completed_sessions + 1
    return {
        "sessions": len(records),
        "hours": round(total_hours, 1),
        "completed": completed_sessions,
    }

weekly_summary = summarize_week(study_records)
print("Weekly summary:", weekly_summary)

Weekly summary: {'sessions': 5, 'hours': 9.0, 'completed': 4}


## Results · Completion Rule Output

The final cell applies the stated threshold to the returned counts and prints both the rate and label.

In [4]:
completion_rate = (
    weekly_summary["completed"] / weekly_summary["sessions"] * 100
)

if completion_rate >= 80:
    completion_label = "steady"
else:
    completion_label = "needs review"

print("Completion rate (%):", round(completion_rate, 1))
print("Completion label:", completion_label)

Completion rate (%): 80.0
Completion label: steady


## Result Analysis

### Applied Theory

Interpret the case with Boolean Rules and Branching, Collections and Iteration, and Functions and Object State.

### Computed Results

The five records total 9.0 hours: Python contributes 4.5 hours, Statistics 2.0 hours, and Data 2.5 hours. Four of five planned sessions are complete, so the calculated completion rate is 80.0% and the stated rule assigns the label `steady`.

### Interpretation

The subject dictionary answers a grouped-total question, while the returned summary dictionary keeps the overall counts and duration labeled. The completion label follows an explicit threshold and does not measure learning quality.

### Assumptions and Limits

The records are synthetic, the durations are treated as accurate, and every session receives equal weight in the completion rate. No conclusion about a real student or causal effect is supported.

### Next Topic

The next session begins with Paths and Filesystem State, which separates location descriptions from observable file effects.
